sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
 from ...utils._typedefs cimport ITYPE_t, DTYPE_t
 from ...utils._typedefs import ITYPE, DTYPE
 
-cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_from_distance_matrix(
-    cnp.ndarray[cnp.double_t, ndim=2] distance_matrix
+# Numpy structured dtype representing a single ordered edge in Prim's algorithm
+MST_edge_dtype = np.dtype([
+    ("current_node", np.intp),
+    ("next_node", np.intp),
+    ("distance", np.float64),
+])
+
+# Packed shouldn't make a difference since they're all 8-byte quantities,
+# but it's included just to be safe.
+ctypedef packed struct MST_edge_t:
+    cnp.intp_t current_node
+    cnp.intp_t next_node
+    cnp.float64_t distance
+
+cpdef cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst_from_mutual_reachability(
+    cnp.ndarray[cnp.float64_t, ndim=2] mutual_reachability
 ):
-
     cdef:
-        cnp.ndarray[cnp.intp_t, ndim=1] node_labels
-        cnp.ndarray[cnp.intp_t, ndim=1] current_labels
-        cnp.ndarray[cnp.double_t, ndim=1] current_distances
-        cnp.ndarray[cnp.double_t, ndim=1] left
-        cnp.ndarray[cnp.double_t, ndim=1] right
-        cnp.ndarray[cnp.double_t, ndim=2] result
-
-        cnp.ndarray label_filter
-
-        cnp.intp_t current_node
-        cnp.intp_t new_node_index
-        cnp.intp_t new_node
-        cnp.intp_t i
-
-    result = np.zeros((distance_matrix.shape[0] - 1, 3))
-    node_labels = np.arange(distance_matrix.shape[0], dtype=np.intp)
+        cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] node_labels
+        cnp.ndarray[cnp.intp_t, ndim=1, mode='c'] current_labels
+        cnp.ndarray[cnp.float64_t, ndim=1, mode='c'] min_reachability, left, right
+        cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst
+
+        cnp.ndarray[cnp.uint8_t, mode='c'] label_filter
+
+        cnp.intp_t n_samples = mutual_reachability.shape[0]
+        cnp.intp_t current_node, new_node_index, new_node, i
+
+    mst = np.empty(n_samples - 1, dtype=MST_edge_dtype)
+    node_labels = np.arange(n_samples, dtype=np.intp)
     current_node = 0
-    current_distances = np.infty * np.ones(distance_matrix.shape[0])
+    min_reachability = np.infty * np.ones(n_samples)
     current_labels = node_labels
-    for i in range(1, node_labels.shape[0]):
+    for i in range(1, n_samples):
```

Comment:
Since the last value of `i` is not used, wouldn't it be better for latex indexing to have:
```suggestion
    for i in range(0, n_samples-1):
```

?